llm-models/embedding/bge/bge-large/04_fine_tune_embedding.py.py

```diff
@@ -0,0 +1,188 @@
+# Databricks notebook source
+# MAGIC %md
+# MAGIC
+# MAGIC # Fine-Tune `bge-large-en` with Sentence Transformers
+# MAGIC
+# MAGIC This notebook demostrates how to fine tune [bge-large-en model](https://huggingface.co/BAAI/bge-large-en).
+# MAGIC
+# MAGIC Environment for this notebook:
+# MAGIC - Runtime: 13.3 GPU ML Runtime
+# MAGIC - Instance: `g5.xlarge` on AWS or `Standard_NV36ads_A10_v5` on Azure.
+
+# COMMAND ----------
+
+import json
+from datasets import load_dataset
+
+# COMMAND ----------
+
+output_model_path = "/dbfs/fine_tuned_bge_model"
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Prepare your dataset for fine tuning
+# MAGIC
+# MAGIC To fine tune a embedding model for document retrieval, the training data requires a label or structure that allows the model to understand whether two sentences are similar or different. One way is to provide a pair of positive (similar) sentences without a label. For example, pairs of paraphrases, pairs of full texts and their summaries, pairs of duplicate questions, pairs of (query, response).
+# MAGIC
+# MAGIC In this tutorial, we will use [Amazon-QA](https://huggingface.co/datasets/embedding-data/Amazon-QA) as an example to build the training dataset.
+
+# COMMAND ----------
+
+from datasets import load_dataset
+
+# Import the data from Huggingface hub.
+dataset = load_dataset("embedding-data/Amazon-QA", split='train[:500]')
+
+# COMMAND ----------
+
+dataset[2]
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC
+# MAGIC You can see that query (the anchor) has a single sentence, pos (positive) is a list of sentences (the one we print has only one sentence). 
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Data preprocessing
+# MAGIC
+# MAGIC Convert the examples into InputExample's.
+
+# COMMAND ----------
+
+from sentence_transformers import InputExample
+
+def create_dataset_for_multiple_loss(train_dataset):
+  train_examples = []
+  for elem in train_dataset:
+    query = f"Represent this sentence for searching relevant passages: {elem['query']}"
+    texts = elem["pos"]
+    
+    for text in texts:
+      train_examples.append(InputExample(texts=[query, text]))
+  return train_example
+
+train_examples = create_dataset_for_multiple_loss(dataset)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC
+# MAGIC ### Fine tune the embedding model
+# MAGIC
+# MAGIC Since we only have sentences pairs with no labels, we can use the `MultipleNegativesRankingLoss` function.
+
+# COMMAND ----------
+
+from sentence_transformers import SentenceTransformer
+model = SentenceTransformer('BAAI/bge-large-en')
+
+# COMMAND ----------
+
+from torch.utils.data import DataLoader
+from sentence_transformers import losses
+from accelerate import notebook_launcher
+
+train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=3)
```

Comment:
Curious about how the batch size is selected?

Reply:
I guess it's just the `batch_size` that doesn't OOM...